content/blog/posts/o3de-motion-matching-2022-04.md

```diff
@@ -0,0 +1,179 @@
+
+The more features selected by the user, the bigger the chances are that the searched and matched pose hits the expected result but the slower the algorithm will be and the more memory it will use. The key is to use crucial and independent elements that define a pose and its movement without being too strict on the wrong end. The root trajectory along with the left and right foot positions and velocities have been proven to be a good start here.
+
+{{< image-width src="/images/blog/motion-matching/feature-matrix.png" width="100%" alt="The motion matching feature matrix." >}}
+
+## Feature matrix
+
+We store the extracted feature values in a giant block of data, the feature matrix. The feature matrix is a NxM matrix that stores the extracted feature values for all frames in our motion database based upon a given feature schema. The feature schema defines the order of the columns and values. It is then used to identify values and find their location inside the matrix.
+
+A 3D position feature storing XYZ values, for example, will use three columns in the feature matrix. Every component of a feature is linked to a column index, so for example, the left foot position Y value might be at column index 6. The group of values or columns that belong to a given feature is what we call a feature block. The accumulated number of dimensions for all features in the schema (while the number of dimensions might vary per feature) form the number of columns of the feature matrix.
+
+Each row represents the features of a single frame of the motion database. As the data is commonly accessed on a per-frame basis, it is stored in row-major order for cache efficiency.
+
+> Memory usage: A motion capture database holding 1 hour of animation data together with a sample rate of 30 Hz to extract features, resulting in 108,000 frames, using the default feature schema having 59 features, will result in a feature matrix holding ~6.4 million values and use ~24.3 MB of memory.
+
+## The motion matching algorithm
+
+Now that we have all the data we need, let's get on to the actual algorithm.
+
+#### Update loop
+
+In the majority of the game ticks, the current motion gets advanced. A few times per second, the actual motion matching search is triggered to prevent drifting too far away from the expected user input (as we would just play the recorded animation otherwise).
+
+When a search for a better next matching frame is triggered, the current pose, including its joint velocities, gets evaluated. This pose (which we'll call input or query pose) is used to fill the query vector. The query vector contains feature values and is compared against other frames in the feature matrix. The query vector has the same size as there are columns in the feature matrix and is similar to any other row but represents the input query.
```

Comment:
Am not quite to sure what you mean by "... and is similar to any other row but represents the input query."

Reply:
Changed to: "The query vector has the same size as there are columns in the feature matrix and is similar to any other row which stores the features for a given frame in the motion database."

Does that make it more clear?